File: test/interop/clients/js-native/tail.ts
/** Finite raw-track interop: publish until the harness acknowledges, or read through clean EOF. @module */
import { strict as assert } from "node:assert";
import { createInterface } from "node:readline";
import * as Moq from "@moq/net";
import { install } from "@moq/web-transport";

install();
const GROUPS = 4;
const BYTES = 256 * 1024;
const [role, url, path] = process.argv.slice(2);
if (!url || !path || !["publish", "subscribe"].includes(role)) {
	throw new Error("usage: tail.ts publish|subscribe URL BROADCAST");
}

const origin = new Moq.Origin.Producer();
const connection = await Moq.Connection.connect({
	url: new URL(url),
	websocket: { enabled: false },
	...(role === "publish" ? { publish: origin.consume() } : { consume: origin }),
});
try {
	if (role === "publish") {
		const broadcast = origin.createBroadcast(Moq.Path.from(path));
		const track = broadcast.createTrack("tail");
		broadcast.announce();
		while (!track.used.peek()) await track.used.changed();
		// The end precedes the last group's data; the transport still has to drain it.
		track.finishAt(GROUPS);
		for (let sequence = 0; sequence < GROUPS; sequence++) {
			const group = track.appendGroup();
			group.writeFrame({
				payload: new Uint8Array(BYTES).fill(sequence),
				timestamp: Moq.Time.Timestamp.fromMillis(0),
			});
			group.close();
		}
		track.close();
		const lines = createInterface({ input: process.stdin });
		try {
			const ack = await lines[Symbol.asyncIterator]().next();
			assert.equal(ack.value, "clean end", "reader did not acknowledge a clean end");
		} finally {
			lines.close();
		}
		broadcast.close();
		console.error("tail acknowledged");
	} else {
		const request = origin.request(Moq.Path.from(path), { announced: true });
		try {
			let broadcast = request.active.peek();
			while (!broadcast) {
				await request.active.changed();
				broadcast = request.active.peek();
			}
			const track = broadcast
				.track("tail")
				.subscribe({ maxAge: Moq.Time.Milli(5000), groups: { start: { included: 0 } } });
			const seen: number[] = [];
			try {
				for (;;) {
					const group = await track.recvGroup();
					if (!group) break;
					try {
						assert(group.sequence < GROUPS, `unexpected group ${group.sequence}`);
						const frame = await group.readFrame();
						assert(frame, "empty group");
						assert.equal(frame.payload.byteLength, BYTES, "truncated group");
						assert(
							frame.payload.every((byte) => byte === group.sequence),
							"corrupt group",
						);
						assert.equal(await group.readFrame(), undefined, "extra frame");
						seen.push(group.sequence);
					} finally {
						group.close();
					}
				}
				assert.deepEqual(
					seen.sort((a, b) => a - b),
					[0, 1, 2, 3],
				);
				assert.equal(await track.finished(), GROUPS);
				console.error("tail clean end=4 groups=0,1,2,3 bytes=1048576");
			} finally {
				track.close();
			}
		} finally {
			request.close();
		}
	}
} finally {
	connection.close();
	origin.close();
}
process.exit(0);

File: test/interop/clients/rust/tail.rs
//! Finite raw-track interop client. The harness acknowledges a clean read on stdin.
use std::time::Duration;

use anyhow::{Context, ensure};
use moq_tokio::moq_net;
use tokio::io::AsyncBufReadExt;

const GROUPS: u64 = 4;
const BYTES: usize = 256 * 1024;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let args: Vec<_> = std::env::args().collect();
    ensure!(
        args.len() == 4,
        "usage: interop-tail publish|subscribe URL BROADCAST"
    );
    let url: url::Url = args[2].parse()?;
    let mut config = moq_tokio::connect::Config::default();
    config.websocket.enabled = Some(false);
    let client = config.init(Default::default())?.with_reconnect(false);
    let origin = moq_tokio::origin::spawn();

    match args[1].as_str() {
        "publish" => {
            let broadcast = origin.create_broadcast(&args[3])?;
            let mut track = broadcast.create_track("tail", None)?;
            broadcast.announce(Default::default())?;
            let _session = client
                .with_publisher(&origin)
                .connect(url)
                .established()
                .await?;
            track.demand().used().await?;
            // Declare the end before the final group's payload is even available.
            track.finish_at(GROUPS)?;
            for sequence in 0..GROUPS {
                let mut group = track.append_group()?;
                group.write_frame(moq_net::Timestamp::ZERO, vec![sequence as u8; BYTES])?;
                group.finish()?;
            }
            // A finished track is not a drained transport. Keep the session alive
            // until the harness confirms the subscriber checked every byte and EOF.
            let mut ack = String::new();
            tokio::io::BufReader::new(tokio::io::stdin())
                .read_line(&mut ack)
                .await?;
            ensure!(
                ack == "clean end\n",
                "reader did not acknowledge a clean end"
            );
            broadcast.finish();
            eprintln!("tail acknowledged");
        }
        "subscribe" => {
            let consumer = origin.consume();
            let _session = client
                .with_subscriber(origin)
                .connect(url)
                .established()
                .await?;
            let broadcast = consumer.routed_broadcast(&args[3]).await?;
            let options = moq_net::track::Subscription::default()
                .with_max_age(Duration::from_secs(5))
                .with_groups(0..);
            let mut track = broadcast.track("tail")?.subscribe(Some(options)).await?;
            let mut seen = Vec::new();
            while let Some(mut group) = track.recv_group().await? {
                let sequence = group.sequence;
                let frame = group.read_frame().await?.context("empty group")?;
                ensure!(sequence < GROUPS, "unexpected group {sequence}");
                ensure!(frame.payload.len() == BYTES, "truncated group {sequence}");
                ensure!(
                    frame.payload.iter().all(|byte| *byte == sequence as u8),
                    "corrupt group {sequence}"
                );
                ensure!(
                    group.read_frame().await?.is_none(),
                    "extra frame in group {sequence}"
                );
                seen.push(sequence);
            }
            seen.sort_unstable();
            ensure!(
                seen == (0..GROUPS).collect::<Vec<_>>(),
                "wrong groups: {seen:?}"
            );
            ensure!(track.finished().await? == GROUPS, "wrong declared end");
            eprintln!("tail clean end=4 groups=0,1,2,3 bytes=1048576");
        }
        _ => anyhow::bail!("unknown role: {}", args[1]),
    }
    Ok(())
}
